### PR TITLE
Add chain ID to member struct

### PR DESCRIPTION
**Pull request by neekolas**
## tl;dr

- Add the chain_id to Members so that we can prevent SCWs from moving between chains
- Serialization and deserialization for the new field

## Next up
- Add validations using the new field
- Add tests for those validations

## Summary by CodeRabbit

- **New Features**
	- Enhanced `AssociationState` and `Member` structures to include optional `chain_id` parameters.
	- Introduced a new error type `AccountIdError` for improved account ID validation.
	- Added optional `added_on_chain_id` field to `Member` struct.

- **Bug Fixes**
	- Improved error handling and validation in serialization processes for member entities.

- **Documentation**
	- Updated configuration for the `neoeinstein-tonic` plugin to include new client module attributes.

- **Tests**
	- Updated test cases to reflect changes in constructors and method signatures for `AssociationState`, `Member`, and `VerifiedSignature`.

### CODE REVIEW

**Comment by neekolas** [comment]
* **#1177** <a href="https://app.graphite.dev/github/pr/xmtp/libxmtp/1177?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @neekolas and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: xmtp_id/src/associations/association_log.rs [hunk 1]
```diff
@@ -183,6 +187,7 @@ impl IdentityAction for AddAssociation {
             new_member_address.clone(),
             Some(existing_entity_id),
             Some(client_timestamp_ns),
+            self.new_member_signature.chain_id,
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/association_log.rs:190]: This is coming in the next PR in the stack

File: xmtp_id/src/associations/mod.rs [hunk 2]
```diff
@@ -312,8 +325,12 @@ pub(crate) mod tests {
...
     async fn reject_invalid_signature_on_create() {
         // Creates a signature with the wrong signer
-        let bad_signature =
-            VerifiedSignature::new(rand_string().into(), SignatureKind::Erc191, rand_vec());
+        let bad_signature = VerifiedSignature::new(
+            rand_string().into(),
+            SignatureKind::Erc191,
+            rand_vec(),
+            None,
+        );
```
[the remark this answers is not in the document]
Comment (neekolas) [reply xmtp_id/src/associations/mod.rs:333]: Coming in the next PR

File: xmtp_id/src/associations/member.rs [hunk 3]
```diff
@@ -78,18 +78,21 @@ pub struct Member {
     pub identifier: MemberIdentifier,
     pub added_by_entity: Option<MemberIdentifier>,
     pub client_timestamp_ns: Option<u64>,
+    pub added_on_chain_id: Option<u64>,
```
Comment (richardhuaaa) [inline xmtp_id/src/associations/member.rs:81]: Just double-checking, do we have a constrained number of chains? E.g., are we protected against someone producing a SCW signature for an address they don't own on a bogus chain? And do revocation signatures have to match the chain on which a SCW was added?

**Review by richardhuaaa — APPROVED** [review]

File: xmtp_id/src/associations/member.rs (hunk 3 above)
Comment (neekolas) [reply xmtp_id/src/associations/member.rs:81]: > do we have a constrained number of chains

Yes. We have a fixed list of chains and to add more we will have to roll the update out to everyone and deprecate old clients.

> And do revocation signatures have to match the chain on which a SCW was added

That PR should be up shortly

**Comment by neekolas** [comment]
### Merge activity

* **Oct 23, 12:27 PM PDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/xmtp/libxmtp/1177).

**Pull request merged**